core/coreapi/dht.go

```diff
@@ -0,0 +1,241 @@
+	ipdht "gx/ipfs/QmY1y2M1aCcVhy8UuTbZJBvuFbegZm47f9cDAdgxiehQfx/go-libp2p-kad-dht"
+	peer "gx/ipfs/QmZoWKhxUmZ2seW4BzX6fJkNR8hh9PsGModr7q171yq2SS/go-libp2p-peer"
+	cid "gx/ipfs/QmcZfnkapfECQGcLZaf9B79NRg7cRa9EnZh4LSbkCzwNvY/go-cid"
+	ipld "gx/ipfs/Qme5bWv7wtjUNGsK2BNGVUFPKiuxWrsqrtvYwCLRw8YFES/go-ipld-format"
+)
+
+var ErrNotDHT = errors.New("routing service is not a DHT")
+
+type DhtAPI CoreAPI
+
+func (api *DhtAPI) FindPeer(ctx context.Context, p peer.ID) (<-chan ma.Multiaddr, error) {
+	dht, ok := api.node.Routing.(*ipdht.IpfsDHT)
+	if !ok {
+		return nil, ErrNotDHT
+	}
+
+	outChan := make(chan ma.Multiaddr)
+	events := make(chan *notif.QueryEvent)
+	ctx = notif.RegisterForQueryEvents(ctx, events)
+
+	go func() {
+		defer close(outChan)
+
+		sendAddrs := func(responses []*pstore.PeerInfo) error {
+			for _, response := range responses {
+				for _, addr := range response.Addrs {
+					select {
+					case outChan <- addr:
+					case <-ctx.Done():
+						return ctx.Err()
+					}
+				}
+			}
+			return nil
+		}
+
+		for event := range events {
+			if event.Type == notif.FinalPeer {
+				err := sendAddrs(event.Responses)
+				if err != nil {
+					return
+				}
+			}
+		}
+	}()
+
+	go func() {
+		defer close(events)
+		pi, err := dht.FindPeer(ctx, peer.ID(p))
+		if err != nil {
+			notif.PublishQueryEvent(ctx, &notif.QueryEvent{
+				Type:  notif.QueryError,
+				Extra: err.Error(),
+			})
+			return
+		}
+
+		notif.PublishQueryEvent(ctx, &notif.QueryEvent{
+			Type:      notif.FinalPeer,
+			Responses: []*pstore.PeerInfo{&pi},
+		})
+	}()
+
+	return outChan, nil
+}
+
+func (api *DhtAPI) FindProviders(ctx context.Context, p coreiface.Path, opts ...caopts.DhtFindProvidersOption) (<-chan peer.ID, error) {
+	settings, err := caopts.DhtFindProvidersOptions(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	dht, ok := api.node.Routing.(*ipdht.IpfsDHT)
+	if !ok {
+		return nil, ErrNotDHT
+	}
+
+	p, err = api.core().ResolvePath(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+
+	c := p.Cid()
+
+	numProviders := settings.NumProviders
+	if numProviders < 1 {
+		return nil, fmt.Errorf("number of providers must be greater than 0")
+	}
+
+	outChan := make(chan peer.ID)
+	events := make(chan *notif.QueryEvent)
+	ctx = notif.RegisterForQueryEvents(ctx, events)
+
+	pchan := dht.FindProvidersAsync(ctx, c, numProviders)
+	go func() {
+		defer close(outChan)
+
+		sendProviders := func(responses []*pstore.PeerInfo) error {
+			for _, response := range responses {
+				select {
+				case outChan <- response.ID:
+				case <-ctx.Done():
+					return ctx.Err()
+				}
+			}
+			return nil
+		}
+
+		for event := range events {
+			if event.Type == notif.Provider {
+				err := sendProviders(event.Responses)
+				if err != nil {
+					return
+				}
+			}
+		}
+	}()
+
+	go func() {
+		defer close(events)
+		for p := range pchan {
+			np := p
+			notif.PublishQueryEvent(ctx, &notif.QueryEvent{
```

Comment:
Again, maybe I'm missing something but this seems like pointless complexity. The caller of this method (`FindProviders`) can't even use these events because we're registering for (and consuming) them.